### ISSUE TITLE
Open components API to allow creating custom components

### ISSUE THREAD

**Issue by malczak** [adobe/react-spectrum#5628]
### Provide a general summary of the feature here

It would be great to be able to use spectrum components as building blocks for custom components. It's already possible to certain degree but still API is closed for extensibility in many places. 

As an example let me present a `MonthPicker` component which is an variation of already available `DatePicker`. The requirement was for user to be able to select a date with month granularity (not possible when using a `DatePicker` component). Picker dropdown had to present months not days. Ideally it should be possible to either extend date picker base component (preferred) or override parts of the `DatePicker` component itself. 

Unfortunately not is currently possible. To build  `MonthPicker` version of the `DatePicker` component its required to make a hard copy of all components building it - i.e. `DatePickerField`, `DatePickerSegment`, `Input`, etc. 

Moreover this requires using a hardcoded css classes with in turn locks us with a specific version of react-spectrum library, making the upgrade expensive and time consuming. 

Imho this could be a blocker for spectrum library to become a widely used UI library - if this is even expected and wanted.

### 🤔 Expected Behavior?

Ideally all base components and all component building blocks, utils etc should be exported from library. 

### 😯 Current Behavior

Currently only official components are exposed publicly from library. Building a custom components lock users to a specific release version (due to hardcoded css classes).

### 💁 Possible Solution

Export all base components and building blocks from library.

### 🔦 Context

While working on an application where react-spectrum was an UI library I had to add a specific (custom) version of available components preserving the look and feel of spectrum library. 

### 💻 Examples

As an example please see my solution to the described problem of creating a custom `MonthPicker` components.
- Working example - https://spectrum-month-picker.vercel.app/
- Source - [spectrum-month-picker](https://github.com/malczak/spectrum-month-picker)

**Comment by reidbarber** [comment]
I think for these use cases, you can use React Aria with Spectrum classes from [Spectrum CSS](https://opensource.adobe.com/spectrum-css/index.html) to get predictable class names, or reference global Spectrum CSS variables.

Alternatively, if you want to use Tailwind, you can check out the [example](https://github.com/adobe/react-spectrum/blob/main/examples/rac-spectrum-tailwind/src/spectrum-preset.js) we have that has a Tailwind preset with Spectrum styles: https://codesandbox.io/p/devbox/silly-zhukovsky-m4s422

We want to maintain alignment with [Spectrum](https://spectrum.adobe.com/) guidelines and components, so I'm not sure we want to export any sub-components, although I understand how they would be beneficial when building similar components. Hopefully these resources can make it easier to build the custom components you need.

**Comment by malczak** [comment]
@reidbarber thanks for bringing up React Aria which is a great library and I've already used for building UI systems in past. I did test Tailwind with Spectrum CSS and it indeed works great. But it wont work for custom component build on top those already existing.

If we are building custom component based on (or similar to) one already existing, we want to keep a look and feel of the source component. In such case there are more cons than pros of using tailwind, aria and spectrum-css as we need to implement already existing private components (e.q. calendar display, date input). This means new bugs, UX differences etc. 

What I mean is that there is a lot of hidden potential and value in an already implemented base code. I do understand your reasoning of staying close to spectrum design guidelines. Opening API is that rare opportunity to eat a cake and have a cake at the same time 😃  I would make is possible to extend react-spectrum and staying aligned with Spectrum guidelines.

**Comment by snowystinger** [comment]
Thanks for the discussion.

Just to clarify, React Spectrum's components are meant to not just align with Spectrum, but are actually intended to enforce it. That is why we have UNSAFE className and style props and also why we export as few components as possible. The month picker is an interesting one, and we may support something to do month selection in the future. I don't know if the official Spectrum Designed month picker will look the same, though, and that is why we don't allow it to be extended at this time.

In the meantime, RAC was developed as a way for users to extend the work we've done to make everything accessible. It doesn't include styles, however, as @reidbarber pointed out, it can be used with Spectrum-css or our tailwind plugin pretty easily. Yes, it's a little more work, but it won't be inadvertently broken by us if we change a style or classname, and it's much more flexible.
In short, what you're trying to do is exactly why we built RAC.

**Comment by malczak** [comment]
@snowystinger @reidbarber  Thank you both for clarifications. It all makes sense when enforcing Spectrum is the design decision.  
RAC is in fact very good solution for custom components. My case described above might simply be too unique.

**Comment by reidbarber** [comment]
Sounds good. I'll close this since we aren't going to be exposing any of these, but feel free to post any follow-up questions here.
